apps/api/src/threads/util/thread-state.ts

```diff
@@ -280,62 +280,79 @@ export async function addAssistantResponse(
+ * `isToolCallFinished` flag to indicate completion status. Sets to false until the final chunk, when it is set to true.
+ *
  *
  * Messages will come in from the LLM or agent as a stream of component
  * decisions, as a flat stream or messages, even though there may be more than
  * one actual message, and each iteration of the message may contain an
  * incomplete tool call.
  *
  * For LLMs, this mostly just looks like a stream of messages that ultimately
- * results in a single final message, so just the last message in the resulting
- * stream has a tool call in it.
+ * results in a single final message.
  *
  * For agents, this may be a stream of multiple distinct messages, (like a user
- * message, then two asisstant messages, then another user message, etc) and we
+ * message, then two assistant messages, then another user message, etc) and we
  * distinguish between them because the `id` of the LegacyComponentDecision will
  * change with each message.
  */
 export async function* fixStreamedToolCalls(
   stream: AsyncIterableIterator<LegacyComponentDecision>,
 ): AsyncIterableIterator<LegacyComponentDecision> {
   let currentDecisionId: string | undefined = undefined;
-
-  let currentToolCallRequest: ToolCallRequest | undefined = undefined;
-  let currentToolCallId: string | undefined = undefined;
   let currentDecision: LegacyComponentDecision | undefined = undefined;
+  let previousToolCallRequest: ToolCallRequest | undefined = undefined;
+  let previousToolCallId: string | undefined = undefined;
 
   for await (const chunk of stream) {
     if (currentDecision?.id && currentDecisionId !== chunk.id) {
-      // we're on to a new chunk, so if we have a previous tool call request, emit it
+      // we're on to a new message, so emit the previous one with complete tool call info
       yield {
         ...currentDecision,
-        toolCallRequest: currentToolCallRequest,
-        toolCallId: currentToolCallId,
+        toolCallRequest: previousToolCallRequest,
+        toolCallId: previousToolCallId,
+        isToolCallFinished: true,
       };
-      // and clear the current tool call request and id
-      currentToolCallRequest = undefined;
-      currentToolCallId = undefined;
+      // and clear the current state
+      previousToolCallRequest = undefined;
+      previousToolCallId = undefined;
     }
 
-    // now emit the next chunk
-    const { toolCallRequest, ...incompleteChunk } = chunk;
-    currentDecision = incompleteChunk;
+    // Track the tool call info from the incoming chunk
     currentDecisionId = chunk.id;
-    currentToolCallId = chunk.toolCallId;
-    currentToolCallRequest = toolCallRequest;
-    yield incompleteChunk;
+    const incomingToolCallId = chunk.toolCallId;
+    const incomingToolCallRequest = chunk.toolCallRequest;
+
+    // Store the tool call info for when we need to emit the final chunk
+    // (either when ID changes or at end of stream)
+    if (incomingToolCallRequest) {
```

Comment:
I don't think you need to check this, because even if this is undefined, you want to set previousToolCallRequest to undefined (i.e. it's more complicated to check then to just do it all the time)

Reply:
Honestly when you do all that simplification it goes back to the original code that was here. I think really the only change to this function that you need is to change `yield incompleteChunk` to `yield { ...chunk, isToolCallFinished: false }`